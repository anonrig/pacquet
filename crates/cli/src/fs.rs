use std::{io, os, path::Path};

#[cfg(unix)]
pub fn symlink_dir<P: AsRef<Path>>(original: P, link: P) -> io::Result<()> {
    os::unix::fs::symlink(original, link)
}

#[cfg(windows)]
pub fn symlink_dir<P: AsRef<Path>>(original: P, link: P) -> io::Result<()> {
    os::windows::fs::symlink_dir(original, link)
}

#[cfg(test)]
pub fn get_filenames_in_folder(path: &Path) -> Vec<String> {
    let mut files = std::fs::read_dir(path)
        .unwrap()
        .map(|e| e.unwrap().file_name().to_string_lossy().to_string())
        .collect::<Vec<_>>();

    files.sort();
    files
}

#[cfg(test)]
pub fn get_all_folders(root: &std::path::PathBuf) -> Vec<String> {
    let mut files = Vec::new();
    for entry in walkdir::WalkDir::new(root) {
        let entry = entry.unwrap();
        let entry_path = entry.path().to_path_buf();
        if entry.file_type().is_dir() || entry.file_type().is_symlink() {
            // We need this mutation to ensure that both Unix and Windows paths resolves the same.
            // TODO: Find a better way to do this?
            let simple_path = entry_path
                .strip_prefix(root)
                .unwrap()
                .components()
                .map(|c| c.as_os_str().to_string_lossy().to_string())
                .collect::<Vec<_>>()
                .join("/");

            if !simple_path.is_empty() {
                files.push(simple_path.as_os_string());
            }
        }
    }
    files.sort();
    files
}


