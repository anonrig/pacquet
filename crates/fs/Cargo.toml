[package]
name                 = "pacquet-fs"
description          = "Filesystem utility functions used by pacquet"
version              = "0.0.1"
publish              = false
authors.workspace    = true
edition.workspace    = true
homepage.workspace   = true
keywords.workspace   = true
license.workspace    = true
repository.workspace = true

[dependencies]
pacquet-task-queue = { workspace = true }

derive_more = { workspace = true }
miette      = { workspace = true }

[target.'cfg(windows)'.dependencies]
junction = { workspace = true }
