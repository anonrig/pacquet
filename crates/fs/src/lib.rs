mod ensure_file;
mod symlink_dir;

pub use ensure_file::*;
pub use symlink_dir::*;

pub mod file_mode;
