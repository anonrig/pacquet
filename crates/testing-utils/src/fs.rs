use std::{io, path::Path};

pub fn get_filenames_in_folder(path: &Path) -> Vec<String> {
    let mut files = std::fs::read_dir(path)
        .unwrap()
        .map(|e| e.unwrap().file_name().to_string_lossy().to_string())
        .collect::<Vec<_>>();

    files.sort();
    files
}

pub fn get_all_folders(root: &std::path::Path) -> Vec<String> {
    let mut files = Vec::new();
    for entry in walkdir::WalkDir::new(root) {
        let entry = entry.unwrap();
        let entry_path = entry.path();
        if entry.file_type().is_dir() || entry.file_type().is_symlink() {
            // We need this mutation to ensure that both Unix and Windows paths resolves the same.
            // TODO: Find a better way to do this?
            let simple_path = entry_path
                .strip_prefix(root)
                .unwrap()
                .components()
                .map(|c| c.as_os_str().to_str().expect("invalid UTF-8"))
                .collect::<Vec<_>>()
                .join("/");

            if !simple_path.is_empty() {
                files.push(simple_path);
            }
        }
    }
    files.sort();
    files
}

// Helper function to check if a path is a symlink or junction
pub fn is_symlink_or_junction(path: &Path) -> io::Result<bool> {
    #[cfg(windows)]
    return junction::exists(&path);

    #[cfg(not(windows))]
    return Ok(path.is_symlink());
}
