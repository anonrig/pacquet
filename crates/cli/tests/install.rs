use assert_cmd::prelude::*;
use command_extra::CommandExtra;
use pacquet_testing_utils::{
    bin::pacquet_with_temp_cwd,
    fs::{get_all_files, get_all_folders, is_symlink_or_junction},
};
use pipe_trait::Pipe;
use std::fs;

#[test]
fn should_install_dependencies() {
    let (command, root, workspace) = pacquet_with_temp_cwd(true);

    eprintln!("Creating package.json...");
    let manifest_path = workspace.join("package.json");
    let package_json_content = serde_json::json!({
        "dependencies": {
            "is-odd": "3.0.1",
        },
        "devDependencies": {
            "fast-decode-uri-component": "1.0.1",
        },
    });
    fs::write(&manifest_path, package_json_content.to_string()).expect("write to package.json");

    eprintln!("Executing command...");
    command.with_arg("install").assert().success();

    eprintln!("Make sure the package is installed");
    assert!(is_symlink_or_junction(&workspace.join("node_modules/is-odd")).unwrap());
    assert!(workspace.join("node_modules/.pacquet/is-odd@3.0.1").exists());

    eprintln!("Make sure it installs direct dependencies");
    assert!(!workspace.join("node_modules/is-number").exists());
    assert!(workspace.join("node_modules/.pacquet/is-number@6.0.0").exists());

    eprintln!("Make sure we install dev-dependencies as well");
    assert!(
        is_symlink_or_junction(&workspace.join("node_modules/fast-decode-uri-component")).unwrap()
    );
    assert!(workspace.join("node_modules/.pacquet/fast-decode-uri-component@1.0.1").is_dir());

    eprintln!("Snapshot");
    let workspace_folders = get_all_folders(&workspace);
    let store_files = get_all_files(&root.path().join("pacquet-store"));
    insta::assert_debug_snapshot!((workspace_folders, store_files));

    drop(root); // cleanup
}

#[test]
fn should_install_exec_files() {
    let (command, root, workspace) = pacquet_with_temp_cwd(true);

    eprintln!("Creating package.json...");
    let manifest_path = workspace.join("package.json");
    let package_json_content = serde_json::json!({
        "dependencies": {
            "pretty-exec": "0.3.10",
        },
    });
    fs::write(&manifest_path, package_json_content.to_string()).expect("write to package.json");

    eprintln!("Executing command...");
    command.with_arg("install").assert().success();

    eprintln!("Listing all files in the store...");
    let store_files = root.path().join("pacquet-store").pipe_as_ref(get_all_files);
    let (exec_files, non_exec_files): (Vec<_>, Vec<_>) =
        store_files.iter().partition(|path| path.ends_with("-exec"));

    eprintln!("Snapshot");
    insta::assert_debug_snapshot!(exec_files);

    #[cfg(unix)]
    {
        use pacquet_testing_utils::fs::is_path_executable;
        use pretty_assertions::assert_eq;

        eprintln!("All files that end with '-exec' are executable, others not");
        assert_eq!(
            store_files.iter().partition(|name| {
                root.path().join("pacquet-store").join(name).pipe_as_ref(is_path_executable)
            }),
            (exec_files, non_exec_files),
        );
    }

    #[cfg(windows)]
    let _ = non_exec_files;

    drop(root); // cleanup
}
