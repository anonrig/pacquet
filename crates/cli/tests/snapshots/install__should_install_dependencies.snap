---
source: crates/cli/tests/install.rs
assertion_line: 51
expression: "(workspace_folders, store_files)"
---
(
    [
        "node_modules",
        "node_modules/.pnpm",
        "node_modules/.pnpm/fast-decode-uri-component@1.0.1",
        "node_modules/.pnpm/fast-decode-uri-component@1.0.1/node_modules",
        "node_modules/.pnpm/fast-decode-uri-component@1.0.1/node_modules/fast-decode-uri-component",
        "node_modules/.pnpm/is-number@6.0.0",
        "node_modules/.pnpm/is-number@6.0.0/node_modules",
        "node_modules/.pnpm/is-number@6.0.0/node_modules/is-number",
        "node_modules/.pnpm/is-odd@3.0.1",
        "node_modules/.pnpm/is-odd@3.0.1/node_modules",
        "node_modules/.pnpm/is-odd@3.0.1/node_modules/is-number",
        "node_modules/.pnpm/is-odd@3.0.1/node_modules/is-odd",
        "node_modules/fast-decode-uri-component",
        "node_modules/is-odd",
    ],
    [
        "v3/files/01/d306b80f4b1678c655f3b0685db1f7a741fc746d251c00e925298ab89630a3167bde8c7d8d779b72770c41f9f302bb1e3f312672ce7aa8e6c156c319eeeada",
        "v3/files/17/54b7271ca0e30dc7d775986711a2b61b298200c12f867bde6f7e19d56047c9ab77a93304520fa1a31b0c3a4e6fb6b310f89dd8cb4d95a5c662807ef733f85f",
        "v3/files/1a/5c01a1321e71a1c263fb83878cbba2686f71f571b42c2d40d93858bf1f4c6ff7f625125a68932c243e75ead4a9d2f18d9b22b3fe55c39fe6605a42b2e70620",
        "v3/files/37/8eb16c230fce354ef7e72722ee708cbefea9a481aa597b06fefc12f665ab60176cbf04d6bc159be42ccdbc51764d71516f86bb28a01b0e60c76a2753f1cf44",
        "v3/files/40/fe18b2c1f241075b732727110323f3e29f9d6c8765ff0222bccfa34784caa8abcd18ffbd4bbfab979e5478ace273a74ed3a0a22f0261334fa7c59e032edbde",
        "v3/files/54/7bcd683e2f9f7b0c2790c637b9231539947be923479534592d91886a56d6262fdc1d8ced24ac2cc3adfda139fe399dc7502d382c7afc536e3057a7cf11e6c0",
        "v3/files/55/dcbc7fff1ac1c4755f4cbf24ba21e0dcae062c802d3db937d2b6182f4554dd85d14529a0c9688eac825a345b63ab8b2e5d7699e2be07178df5bb8a072235eb",
        "v3/files/58/01e592e94aa4076c179180956f33f8262fa7157a06bfe636a6854bd3b9b2ddf44780ffa1405642ad8ece2ed6514e19a28178cbecdcf5608a3be710bbf3c0d4",
        "v3/files/60/ab54849d6cfd7b05076c3bc4d9baffe1f800f4d5e2d67c39c521a7a3d93b347ce5a8e45397102e40e2103652c543c8e7bb161c15d0a6e34ec1a9d11c168d2a",
        "v3/files/88/1f101d2f1886fc5f9ac5c8d25e5540908e140031169ea41a4a9941cf95fe2711c1292c1995609c6f174b3815cc0dcc6cd2a10339a348d39fe75bc77b53ba9e",
        "v3/files/9f/cbe5dc86bc33bdb4f125559e767c3aaad94c7c4b21b80b694d9b0158ee0f10ee53b127c2bf05db9768afdaf1fb6778fd08a19412fb3c1adad655863e6c351d",
        "v3/files/ca/65df16eb450ab63f90f107a8cc9ac82da9635408e813a6f6d3e13cfba78917b31b66652479ebba54b612566600b2ab1bc16b82d082c5f87803961a8e9e3931",
        "v3/files/e5/c2514d2b0d3d6bedad24fdd75c55e780211cf0692c21f401201c2d7dfa990ad00188bd68b30c826aaa2a72ecad12a74bc03430384f933b1751035dd1b18ca2",
        "v3/files/f7/02180f4465a777127ad6a211306dd564993c17f7d3d9401604df4bc657d8d888902de632a4cf55b551ddfb9f84ba8a900788773f03c77836143ef94e182c7c",
        "v3/files/ff/d3f3fed65d7b489216916772c6180b5c1714ddde16e7e2c99ff5f8a4214bb8f884a0c1d06a1b1a8a4b0d6252ad85b533b1626f0dd8d474f2283f6cdfd69f6c",
    ],
)
